front/lib/api/assistant/conversation.ts

```diff
@@ -7,14 +7,133 @@ import {
 import { Authenticator } from "@app/lib/auth";
+import { CoreAPI } from "@app/lib/core_api";
+import { Err, Ok, Result } from "@app/lib/result";
+import { isRetrievalActionType } from "@app/types/assistant/actions/retrieval";
 import {
   AssistantAgentMessageType,
   AssistantConversationType,
   AssistantMention,
   AssistantUserMessageContext,
   AssistantUserMessageType,
+  isAgentMessageType,
+  isUserMessageType,
 } from "@app/types/assistant/conversation";
 
+import { renderRetrievalActionForModel } from "./actions/retrieval";
+
+/**
+ * Model rendering of conversations.
+ */
+
+export type ModelMessageType = {
+  role: "action" | "agent" | "user";
+  name: string;
+  content: string;
+};
+
+export type ModelConversationType = {
+  messages: ModelMessageType[];
+};
+
+// This function transforms a conversation in a simplified format that we feed the model as context.
+// It takes care of truncating the conversation all the way to `allowedTokenCount` tokens.
+export async function renderConversationForModel({
+  conversation,
+  model,
+  allowedTokenCount,
+}: {
+  conversation: AssistantConversationType;
+  model: { providerId: string; modelId: string };
+  allowedTokenCount: number;
+}): Promise<Result<ModelConversationType, Error>> {
+  const messages = [];
+
+  let retrievalFound = false;
+
+  // Render all messages and all actions but only keep the latest retrieval action.
+  for (let i = conversation.content.length - 1; i >= 0; i--) {
+    const versions = conversation.content[i];
+    const m = versions[versions.length - 1];
+
+    if (isAgentMessageType(m)) {
+      if (m.action) {
+        if (isRetrievalActionType(m.action) && !retrievalFound) {
+          messages.unshift(renderRetrievalActionForModel(m.action));
+          retrievalFound = true;
+        } else {
+          return new Err(
+            new Error(
+              "Unsupported action type during conversation model rendering"
+            )
+          );
+        }
+      }
+      if (m.message) {
+        messages.unshift({
+          role: "agent" as const,
+          name: m.configuration.name,
+          content: m.message,
+        });
+      }
+    }
+    if (isUserMessageType(m)) {
+      messages.unshift({
+        role: "user" as const,
+        name: m.context.username,
+        content: m.message,
+      });
+    }
+  }
+
+  async function tokenCountForMessage(
+    message: ModelMessageType,
+    model: { providerId: string; modelId: string }
+  ): Promise<Result<number, Error>> {
+    const res = await CoreAPI.tokenize({
+      text: message.content,
+      providerId: model.providerId,
+      modelId: model.modelId,
+    });
+
+    if (res.isErr()) {
+      return new Err(new Error(`Error tokenizing model message: ${res.error}`));
+    }
+
+    return new Ok(res.value.tokens.length);
+  }
+
+  // This is a bit aggressive but fuck it.
+  const tokenCountRes = await Promise.all(
+    messages.map((m) => {
+      return tokenCountForMessage(m, model);
```

Comment:
let's monitor how it goes, but may be useful at some point to have a dedicated thread pool in `core` for tokenizing, right now it creates a new `spawn_blocking` thread each time which is suboptimal

Reply:
Yes! We have 32 workers right now but definitely agree we should move it in its own thread pool :+1: 

Creating an issue on the Assistant board

Reply:
https://github.com/dust-tt/dust/issues/1279